src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs

```diff
@@ -387,5 +387,67 @@ internal static bool IsOutVarDeclaration(this DeclarationExpressionSyntax p)
+        /// <param name="action"></param>
+        /// <param name="argument">The argument that is passed to the action whenever it is invoked</param>
+        internal static void VisitRankSpecifiers<TArg>(this TypeSyntax type, Action<ArrayRankSpecifierSyntax, TArg> action, TArg argument)
+        {
+            switch (type.Kind())
+            {
+                case SyntaxKind.ArrayType:
+                    var arrayTypeSyntax = (ArrayTypeSyntax)type;
+                    arrayTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    foreach (var rankSpecifier in arrayTypeSyntax.RankSpecifiers)
+                    {
+                        action(rankSpecifier, argument);
+                    }
+                    break;
+                case SyntaxKind.NullableType:
+                    var nullableTypeSyntax = (NullableTypeSyntax)type;
+                    nullableTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    break;
+                case SyntaxKind.PointerType:
+                    var pointerTypeSyntax = (PointerTypeSyntax)type;
+                    pointerTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    break;
+                case SyntaxKind.TupleType:
+                    var tupleTypeSyntax = (TupleTypeSyntax)type;
+                    foreach (var element in tupleTypeSyntax.Elements)
+                    {
+                        element.Type.VisitRankSpecifiers(action, argument);
+                    }
+                    break;
+                case SyntaxKind.RefType:
+                    var refTypeSyntax = (RefTypeSyntax)type;
+                    refTypeSyntax.Type.VisitRankSpecifiers(action, argument);
+                    break;
+                case SyntaxKind.GenericName:
+                    var genericNameSyntax = (GenericNameSyntax)type;
+                    foreach (var typeArgument in genericNameSyntax.TypeArgumentList.Arguments)
+                    {
+                        typeArgument.VisitRankSpecifiers(action, argument);
+                    }
+                    break;
+                case SyntaxKind.QualifiedName:
+                    var qualifiedNameSyntax = (QualifiedNameSyntax)type;
+                    qualifiedNameSyntax.Left.VisitRankSpecifiers(action, argument);
+                    qualifiedNameSyntax.Right.VisitRankSpecifiers(action, argument);
+                    break;
+                case SyntaxKind.AliasQualifiedName:
+                    var aliasQualifiedNameSyntax = (AliasQualifiedNameSyntax)type;
+                    aliasQualifiedNameSyntax.Name.VisitRankSpecifiers(action, argument);
+                    break;
+                case SyntaxKind.IdentifierName:
+                case SyntaxKind.OmittedTypeArgument:
+                case SyntaxKind.PredefinedType:
+                    break;
+                default:
+                    throw ExceptionUtilities.Unreachable;
```

Comment:
>throw ExceptionUtilities.Unreachable; [](start = 20, length = 37)

Since we are switching on the Kind, it might be better to ```UnexpectedValue```` helper instead of ```Unreachable```.